python/cudf/cudf/_lib/io/utils.pyx

```diff
@@ -27,3 +31,35 @@ cdef source_info make_source_info(src) except*:
     else:
         raise TypeError("Unrecognized input type: {}".format(type(src)))
     return source_info(<char*>&buf[0], buf.shape[0])
+
+
+# Converts the Python sink input to libcudf++ IO sink_info.
+cdef sink_info make_sink_info(src, unique_ptr[data_sink] * data) except*:
+    if isinstance(src, io.IOBase):
+        data.reset(new iobase_data_sink(src))
+        return sink_info(data.get())
+    elif isinstance(src, (int, float, complex, basestring, os.PathLike)):
+        return sink_info(<string> str(src).encode())
+    else:
+        raise TypeError("Unrecognized input type: {}".format(type(src)))
+
+
+# Adapts a python io.IOBase object as a libcudf++ IO data_sink. This lets you
+# write from cudf to any python file-like object (File/BytesIO/SocketIO etc)
+cdef cppclass iobase_data_sink(data_sink):
+    object buf
+
+    iobase_data_sink(object buf_):
+        this.buf = buf_
+
+    void host_write(const void * data, size_t size) nogil:
+        with gil:
+            buf.write(PyMemoryView_FromMemory(<char*>data, size, PyBUF_READ))
+
+    void flush() nogil:
+        with gil:
+            buf.flush()
+
+    size_t bytes_written() nogil:
+        with gil:
+            return buf.tell()
```

Comment:
Any reason to mark these as `nogil` and then all of the functions need the gil so they use a context `with gil:`?

Reply:
Did you try just doing?
```
void flush() with gil:
    buf.flush()
```

This just feels no bueno that we could be thrashing releasing and reacquiring the GIL unnecessarily.

Reply:
I hadn't tried that - and it works great, changed in latest commit. Thanks!